lookout/style/format/annotations/annotated_data.py

```diff
@@ -0,0 +1,265 @@
+        self._doc_start_offset = numpy.array(doc_lens).cumsum()
+
+    def _offset_to_doc_index(self, offset_start: int, offset_stop: int) -> Tuple[int, int, int]:
+        if offset_start >= len(self):
+            raise IndexError("Start offset %d is greater then the collection length %d." % (
+                offset_start, len(self)))
+        if offset_stop > len(self):
+            raise IndexError("Stop offset %d is greater then the collection length %d." % (
+                offset_stop, len(self)))
+
+        doc_index = numpy.argmax(self._doc_start_offset > offset_start) - 1
+        doc_offset_start = offset_start - self._doc_start_offset[doc_index]
+        doc_offset_stop = offset_stop - self._doc_start_offset[doc_index]
+        if doc_offset_stop > len(self._document_collection[doc_index]):
+            raise IndexError("You can get data only from one document from collection")
+
+        return doc_index, doc_offset_start, doc_offset_stop
+
+    def __len__(self):
+        """Accumulative document collection length."""
+        return self._doc_start_offset[-1]
+
+    def __getitem__(self, index: Union[slice, int, tuple]) -> str:
+        """
+        Access data slices by offset.
+        """
+        if isinstance(index, int):
+            doc_index, doc_offset = self._offset_to_doc_index(index)
+            return self._document_collection[doc_index][doc_offset]
+        elif isinstance(index, slice):
+            if isinstance(index, slice) and index.step is not None:
+                raise IndexError("Step in unsupported for slices.")
+            return self._get_range(index.start, index.stop)
+        elif isinstance(index, tuple):
+            assert len(index) == 2
+            return self._get_range(*index)
+        else:
+            raise IndexError("Unknown index type %s" % type(index))
+
+    def _get_range(self, start: int, stop: int) -> str:
+        if start == stop:
+            return ""
+        doc_index_start, doc_offset_start, doc_offset_end = self._offset_to_doc_index(start, stop)
+        return self._document_collection[doc_index_start][doc_offset_start: doc_offset_end]
+
+    def get_docs_range(self):
+        """Get documents index range."""
+        return tuple(zip(self._doc_start_offset, self._doc_start_offset[1:]))
+
+
+class AnnotatedData:
+    """
+    Class that couples annotations and data together.
+
+    All special utilities to work with annotations should be implemented in this class
+    List of methods that should be implemented can be found here:
+    https://uima.apache.org/d/uimafit-current/api/org/apache/uima/fit/util/JCasUtil.html
+    """
+
+    def __init__(self, raw_data: RawData):
+        """Init."""
+        self._raw_data = raw_data
+        # Interval trees should be used for _range_to_annotations later.
+        self._range_to_annotations = SortedDict()  # type: SortedDict[(int, int), Dict[str, Annotation]]  # noqa E501
```

Comment:
Why do we need to sort?

Reply:
Because it is possible to add annotations in any order. And I want to have a fast way to iterate through annotations.

Reply:
Yep, but why do we care about this order? We can iterate in any order and it should not matter, should it

Reply:
The most common use-case for (Ordered) List of VirtualNodes we have right now is exactly iterating through it in the correct order. We are going to replace this List with Annotations and I want to be able to iterate fast trough it, so I use ordered Dict for that. I do not need to sort to iterate through it. 

It also allows us to search for overlapping/covered/equal annotations by log time. This is also a common usecase.

Reply:
Thanks for explaining

Reply:
May be worth putting this as a comment in the code